File: ruby/qd.rb
# The more you know.
# https://gist.github.com/olegispe/ab80a55735b1652b3618dad85e04551e

require 'io/console'

class Interpreter
  def initialize code
    @stack = []
    @direction = :E
    @running = true
    @x = -1
    @y = 0
    @goto = [0,0]

    @map = []
    rows = code.delete(' ').split("\n")
    # For each row, get columns
    rows.each { |row|
      if not row.match(/^#/)
        @map.push row.split ""
      end
    }
  end

  def execute
    p @map
    begin
      # sleep 0.1
      move
      @token = @map[@y][@x]
      # puts "#{@token} - #{@x}, #{@y}, [#{@stack.join(', ')}]"
      run
    end while @running

    puts "\n\nFinished with:"
    puts "[" + @stack.join(", ") + "]"
  end
  
  def move
    case @direction
    when :N
      @y -= 1
    when :NE
      @x += 1
      @y -= 1
    when :E
      @x += 1
    when :SE 
      @x += 1
      @y += 1
    when :S 
      @y += 1
    when :SW 
      @x -= 1
      @y += 1
    when :W 
      @x -= 1
    when :NW 
      @x -= 1
      @y -= 1
    end

    if @x < 0 or @y < 0 or @x > @map[0].length or @y > @map.length
      puts "Flat-earth error #{@x},#{@y}"
      @running = false
    end
  end

  def run
    # These will be returned
    case @token
    when "?"
      char = STDIN.getch
      if char == "\u0003"
        @running = false
      end
      @stack.push char.ord
    when "!"
      char = @stack.pop
      print char.chr
    when "#"
      a = @stack.pop
      b = @stack.pop
  
      @stack.push b
      @stack.push a
    when "&"
      item = @stack.pop
      2.times {
        @stack.push item
      }
    when "="
      a = @stack.pop
      b = @stack.pop
  
      if a == b
        @stack.push 1
      else
        @stack.push 0
      end
    when "0", "1", "2", "3", "4", "5", "6", "7", "8", "9"
      @stack.push @token.to_i
    when "+"
      a = @stack.pop
      b = @stack.pop
      
      if a + b >= 256
        @stack.push 0
      else
        @stack.push a + b
      end
    when "-"
      a = @stack.pop
      b = @stack.pop
  
      @stack.push b - a
    when "*"
      a = @stack.pop
      b = @stack.pop
  
      @stack.push b * a
    when ":"
      a = @stack.pop
      b = @stack.pop
  
      @stack.push b / a
    when "$"
      @stack.pop
    when "\""
      @stack.reverse!
    when "'"
      @stack.push @stack.pop
    when "~"
      if @stack.pop == 0
        move
      end
    when ","
      @goto = [@x, @y]
    when ";"
      place = @goto
      @x = place[0]
      @y = place[1]
    when "@"
      @running = false
    # Directions
    when "^"
      @direction = :N
    when ">"
      @direction = :E 
    when "v"
      @direction = :S
    when "<"
      @direction = :W
    when "/"
      case @direction
      when :N, :W
        @direction = :SE
      when :S, :E
        @direction = :NW
      end
    when "\\"
      case @direction
      when :N, :E
        @direction = :SW
      when :S, :W 
        @direction = :NE 
      end
    when "_"
      case @direction
      when :N
        @direction = :S
      when :NE 
        @direction = :SE
      when :NW
        @direction = :SW
      when :S 
        @direction = :N
      when :SE 
        @direction = :NE
      when :SW
        @direction = :NW
      end
    when "|"
      case @direction
      when :E
        @direction = :W
      when :NE 
        @direction = :NW
      when :NW
        @direction = :NE
      when :W
        @direction = :E
      when :SE 
        @direction = :SW
      when :SW
        @direction = :SE
      end
    end
  end
end

(Interpreter.new File.read(ARGV[0])).execute

File: js/qd.js
function print (char) {
  // Print char to output box.
  document.getElementById('out').value += char
}

function sleep(milliseconds) {
  var start = new Date().getTime();
  while (!(new Date().getTime() - start) > milliseconds) {}
}

var INTERVAL = 100

function Interpreter (code) {
  this.stack = []
  this.token = ''
  this.direction = 'E'
  this.running = true
  this.x = -1
  this.y = 0
  this.goto = [0, 0]

  this.map = []
  var rows = code.replace(/\s*/g, '').split(/\r|\r\n|\n/)

  rows.forEach(row => {
    if (!row.startsWith('#')) {
      this.map.push(row.split(''))
    }
  })
}

Interpreter.prototype.execute = function () {
  this.move()
  this.token = this.map[this.y][this.x]
  document.getElementById('char').innerText = `${this.token} [${this.stack.join(', ')}]`
  this.run()
  var _this = this
  if (this.running) {
    setTimeout(function () { _this.execute() }, parseInt(document.getElementById('interval').value))
  }
}

Interpreter.prototype.move = function () {
  switch (this.direction) {
    case 'N':
      this.y -= 1
      break
    case 'NE':
      this.x += 1
      this.y -= 1
      break
    case 'E':
      this.x += 1
      break
    case 'SE':
      this.x += 1
      this.y += 1
      break
    case 'S':
      this.y += 1
      break
    case 'SW':
      this.x -= 1
      this.y += 1
      break
    case 'W':
      this.x -= 1
      break
    case 'NW':
      this.x -= 1
      this.y -= 1
      break
    default:
      break
  }

  if (this.x < 0 || this.y < 0 || this.x > this.map[0].length || this.y > this.map.length) {
    console.log('Fell off edge.')
    this.running = false
  }
}

Interpreter.prototype.run = function () {
  var char
  var item
  var a
  var b
  switch (this.token) {
    case '?':
      char = prompt('Input: ').charCodeAt(0)

      this.stack.push(char)
      break
    case '!':
      char = this.stack.pop()
      print(String.fromCharCode(char))
      break
    case '#':
      a = this.stack.pop()
      b = this.stack.pop()

      this.stack.push(b)
      this.stack.push(a)
      break
    case '&':
      item = this.stack.pop()

      this.stack.push(item)
      this.stack.push(item)
      break
    case '=':
      a = this.stack.pop()
      b = this.stack.pop()

      if (a === b) {
        this.stack.push(1)
      } else {
        this.stack.push(0)
      }
      break
    case '0':
    case '1':
    case '2':
    case '3':
    case '4':
    case '5':
    case '6':
    case '7':
    case '8':
    case '9':
      this.stack.push(parseInt(this.token))
      break
    case '+':
      a = this.stack.pop()
      b = this.stack.pop()

      this.stack.push(a + b)
      break
    case '-':
      a = this.stack.pop()
      b = this.stack.pop()

      this.stack.push(b - a)
      break
    case '*':
      a = this.stack.pop()
      b = this.stack.pop()

      this.stack.push(b * a)
      break
    case ':':
      a = this.stack.pop()
      b = this.stack.pop()

      this.stack.push(b / a)
      break
    case '$':
      this.stack.pop()
      break
    case '"':
      this.stack = this.stack.reverse()
      break
    case '~':
      if (this.stack.pop() === 0) {
        this.move()
      }
      break
    case ',':
      this.goto = [this.x, this.y]
      break
    case ';':
      this.x = this.goto[0]
      this.y = this.goto[1]
      break
    case '@':
      this.running = false
      break
    case '^':
      this.direction = 'N'
      break
    case '>':
      this.direction = 'E'
      break
    case 'v':
      this.direction = 'S'
      break
    case '<':
      this.direction = 'W'
      break
    case '\\':
      switch (this.direction) {
        case 'N':
        case 'E':
          this.direction = 'SE'
          break
        case 'S':
        case 'W':
          this.direction = 'NW'
          break
        default:
          break
      }
      break
    case '/':
      switch (this.direction) {
        case 'N':
        case 'E':
          this.direction = 'SW'
          break
        case 'S':
        case 'W':
          this.direction = 'NE'
          break
        default:
          break
      }
      break
    case '_':
      switch (this.direction) {
        case 'N':
          this.direction = 'S'
          break
        case 'NE':
          this.direction = 'SE'
          break
        case 'NW':
          this.direction = 'SW'
          break
        case 'S':
          this.direction = 'N'
          break
        case 'SE':
          this.direction = 'NE'
          break
        case 'SW':
          this.direction = 'NW'
          break
        default:
          break
      }
      break
    case '|':
      switch (this.direction) {
        case 'E':
          this.direction = 'W'
          break
        case 'NE':
          this.direction = 'NW'
          break
        case 'NW':
          this.direction = 'NE'
          break
        case 'W':
          this.direction = 'E'
          break
        case 'SE':
          this.direction = 'SW'
          break
        case 'SW':
          this.direction = 'SE'
          break
        default:
          break
      }
      break
    default:
      break
  }
}
